canals/component/input_output.py

```diff
@@ -2,78 +2,151 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 import logging
-import inspect
-from dataclasses import fields
+from enum import Enum
+from dataclasses import fields, is_dataclass, dataclass, asdict, MISSING
+
+from canals.errors import ComponentError
 
 logger = logging.getLogger(__name__)
 
 
-class BaseIODataclass:  # pylint: disable=too-few-public-methods
+def _make_fields_optional(class_: type):
     """
-    Base class for input and output classes of components.
+    Takes a dataclass definition and modifies its __init__ so that all fields have
+    a default value set.
+    If a field has a default factory use it to set the default value.
+    If a field has neither a default factory or value default to None.
     """
+    defaults = []
+    for field in fields(class_):
+        default = field.default
+        if field.default is MISSING and field.default_factory is MISSING:
+            default = None
+        elif field.default is MISSING and field.default_factory is not MISSING:
+            default = field.default_factory()
+        defaults.append(default)
+    # mypy complains we're accessing __init__ on an instance but it's not in reality.
+    # class_ is a class definition and not an instance of it, so we're good.
+    # Also only I/O dataclasses are meant to be passed to this function making it a bit safer.
+    class_.__init__.__defaults__ = tuple(defaults)  # type: ignore
+
+
+def _make_comparable(class_: type):
+    """
+    Overwrites the existing __eq__ method of class_ with a custom one.
+    This is meant to be used only in I/O dataclasses, it takes into account
+    whether the fields are marked as comparable or not.
 
-    def names(self):
-        """
-        Returns the name of all the fields of this dataclass.
-        """
-        return [field.name for field in fields(self)]
-
+    This is necessary since the automatically created __eq__ method in dataclasses
+    also verifies the type of the class. That causes it to fail if the I/O dataclass
+    is returned by a function.
 
-class Optionalize(type):
-    """
-    Makes all the fields of the dataclass optional by setting None as their default value.
+    In here we don't compare the types of self and other but only their fields.
     """
 
-    def __call__(cls, *args, **kwargs):
-        obj = cls.__new__(cls, *args, **kwargs)
-        obj.__init__.__func__.__defaults__ = tuple(None for _ in inspect.signature(obj.__init__).parameters)
-        obj.__init__(*args, **kwargs)
-        return obj
+    def comparator(self, other) -> bool:
+        if not is_dataclass(other):
+            return False
 
+        # We don't care about the order, if they're the same class
+        # the fields will have the same order
```

Comment:
Eeeh, I wouldn't. Dataclasses' fields are stored in the order they're declared, order is guaranted.

Reply:
Yep what I mean is, are we sure the order in which they're declared is guaranteed? If the class is created dynamically depending on the input, it might not. Consider this code:

```python
@component

class MyComponent

    @component.input
    def input(self):
        return make_dataclass("Input", [category, str, None] for category in self.categories)
    
    ...

    def __init__(self, categories: List[str]):
        self.categories = categories

    ...

c1 = MyComponent(categories=["txt", "pdf", "doc"])
c2 = MyComponent(categories=["txt", "pdf", "doc"])
c3 = MyComponent(categories=["doc", "pdf", "txt"])

assert c1.input() == c2.input()
assert c1.input() == c3.input()
```

The first passes, while the second fails. Looks a bit odd. Is that intended? Technically the order should not matter...